#Install the package when you first use this script
#if (!require("BiocManager", quietly = TRUE))
  #install.packages("BiocManager")

#BiocManager::install("DESeq2")

remove(list=ls())
gc()
# This script uses DESeq2 in R to run differential-expression analysis on a reads count matrix of genes and automatically generate some graphs.
# Load libraries
library(DESeq2)
library(RColorBrewer)
library(calibrate)
library(ggplot2)
library(limma)

# Read in the data and make sample information table
# read the reads count matrix file that was generated by rsem-generate-data-matrix
featureCounts <- read.delim("D:/TDP43/06202023/SRRMETADATA/NEW YORK/featureCounts.txt", comment.char="#")				
counts <-featureCounts[,7:75]
rownames(counts) <- featureCounts[,1]
# get the number of columns
cols=c(1:ncol(counts))
# format the matrix as integer numbers for DESeq2 to be happy
cts<-as.matrix(counts)
# make samples list for DESeq2
samples <- data.frame(colnames(counts))
samples$batch1 <- c("3","4","1","2","3","1","3","1","2",
                    "2","4","3","2","2","1","4","1","2",
                    "3","4","4","2","3","3","3","1","4","3",
                    "3","4","4","4","4","4","3","3","4","3",
                    "2","2","3","1","3","2","1","2","1","4",
                    "2","4","2","3","3","2","2","4","1","1",
                    "4","4","2","2","1","1","1","3","1","3",
                    "1")
samples$gender <- c("M","M","F","F","M","F","M","F","F",
                   "F","M","M","F","F","F","M","F","F",
                   "F","M","M","F","M","M","M","F","M","M",
                   "M","M","M","M","M","M","M","M","M","M",
                   "F","M","F","M","F","F","M","F","F","F",
                   "F","M","F","M","M","F","F","M","F","F",
                   "M","M","F","F","F","F","M","F","M","F",
                   "M")
#1-NEU 2-PF-UCL 3-XX-ctrl-xx 4-mssm-jc-xx 5-xx-xxx-xx
samples$projectid <- c("5","5","5","5","5","5","5","5","1",
                      "1","1","1","1","1","1","1","1","1",
                      "2","1","1","1","3","3","3","4","4","3",
                      "3","4","2","2","2","2","3","4","4","3",
                      "1","2","2","2","2","1","1","2","1","1",
                      "1","2","2","2","1","1","1","1","2","2",
                      "1","2","2","2","2","2","2","2","2","2",
                      "2")

samples$batch1 <- factor(samples$batch1,levels = c("1","2","3","4"))
samples$gender <- factor(samples$gender,levels=c("M","F"))
samples$projectid <- factor(samples$projectid ,levels = c("1","2","3","4","5"))
#Condition of ALS
samples$condition <- c(rep(c("ALS"),18),rep(c("Control"),51))
samples=as.data.frame(row.names = (colnames(counts)),samples)

#convert condition and gender to factor
samples$condition<- factor(samples$condition,levels = c("Control","ALS"))

# Run DESeq2 to get the result for differential-expression
dds <- DESeqDataSetFromMatrix(countData = cts,
                              colData = samples,
                              design= ~ condition,
                              )				# generate a DESeq2 object from the data
 #set the reference level
dds$condition <- relevel(dds$condition,ref="Control")    
colnames(dds) <- colnames(cts)
#trans names
library(AnnotationDbi)
library(org.Hs.eg.db)
library(biomaRt)
ensamble=gsub("\\..*","",rownames(counts))
ensamble <- rownames(counts)
SYMBOL=mapIds(org.Hs.eg.db,
                      key=ensamble,
                      column = "SYMBOL",
                      keytype = "ENSEMBL",
                      multiVals = "first"     )
symbol = as.vector(unlist(SYMBOL))
ids = data.frame(counts,symbol,stringsAsFactors = F)
ids=ids[ !is.na(ids$symbol),]
#####################################
library(PCAtools)
vst<- assay(vst(dds))
p <- pca(vst, metadata =samples)
screeplot(p, axisLabSize = 18, titleLabSize = 22)
biplot(p)

#PCA plot
#log transfer
vsd<- vst(dds)
plotPCA(vsd, intgroup=c("condition","projectid"))+geom_text(aes(label=rownames(samples)),size=3)
plotPCA(vsd, intgroup=c("gender","projectid"))+geom_text(aes(label=rownames(samples)),size=3)
#
assay(vsd) <- limma::removeBatchEffect(assay(vsd), samples$gender)
assay(vsd) <- limma::removeBatchEffect(assay(vsd), samples$projectid)
plotPCA(vsd, intgroup=c("condition","projectid"))+geom_text(aes(label=rownames(samples)),size=3)
p1 <- pca(vst1, metadata =samples)
biplot(p1)
vst2 <- limma::removeBatchEffect(vst1, samples$batch2)
p2 <- pca(vst2, metadata =samples)
biplot(p2)


####################################		
# run DESeq2
dds <- DESeq(dds)
nrow(dds)
#filter out low count data
dds =dds[rowSums(counts(dds))>10,]
nrow(dds)
# Save differential expression results
res <- results(dds)
table(res$padj<=0.05)
## Order by adjusted p-value
res <- res[order(res$padj), ]
## Merge with normalized count data
allcounts<- as.data.frame(counts(dds, normalized=TRUE))
resdata <- merge(as.data.frame(res),allcounts, by="row.names", sort=FALSE)
rownames(resdata) <- resdata$Row.names
signi<- subset(resdata,padj<=0.05)
##################################



library(dplyr)
a<- filter(resdata,!is.na(SYMBOL))
## Write results
write.csv(resdata, file=paste0(output, cond1, "_vs_", cond2, "-ExpDiff.csv"))






##volcano plot
res1 <- as.data.frame(res)
res1$group <- ifelse(res1$padj<0.05 & abs(res1$log2FoldChange)>= 1,
                      ifelse(res1$log2FoldChange > 1,'up','down'),'none')
p <- ggplot(
  res1, aes(log2FoldChange, -log10(padj), col = group)) +  
  geom_point() +
  theme_bw() +
  scale_color_manual(values = c("#FF0000A0", "#A0A0A0A0", "#0000FFA0"), limits = c('up', 'none', 'down')) +
  labs(x="log2 Fold Change",y="-log10 adjust p-value", title = "Whole Blood") +
  geom_hline(yintercept = -log10(0.05), lty=4,col="grey",lwd=0.6) +
  geom_vline(xintercept = c(-1, 1), lty=4,col="grey",lwd=0.6) +
  ylim(0,5)+
  theme(legend.position = "none",
        panel.grid=element_blank(),
        axis.title = element_text(size = 16),
        axis.text = element_text(size = 14)
  )
p



##

